### DISCUSSION TITLE
I need guidance!

### DISCUSSION THREAD

**Discussion by Divaryab** [airsdk/Adobe-Runtime-Support#834] [Q&A] [answered]
Hi All,

 First, My work environment is:
  - **Windows 10** (64 bit)
  - SWF design and compilation in **Adobe Flash CS6** (I did not update to Adobe Animate due to TLFText)
  - Coding in **FlashDevelop** (and running on Flash CS6 by hitting F6)
  - Use the latest version of the SDK (**33.1.1.476**)
  - Use the latest version of ANEs (from distriqt)
  - The manifest is something like this:
  
  ```
<?xml version="1.0" encoding="UTF-8" standalone="no" ?>
<application 
  xmlns="http://ns.adobe.com/air/application/32.0"
  minimumPatchLevel="0">
	<id>org.epayam.QuranHakim</id>
	<versionLabel>7.4</versionLabel>  
	<versionNumber>7.3.23</versionNumber>
	<filename>QuranHakim</filename>
	<description/>	
	<name>Quran Hakim</name>	
	<copyright/>
	<initialWindow>
		<content>QH.swf</content>
		<systemChrome>standard</systemChrome>
		<transparent>false</transparent>
		<visible>true</visible>
		<fullScreen>true</fullScreen>
		<aspectRatio>portrait</aspectRatio>
		<renderMode>cpu</renderMode>
		<autoOrients>false</autoOrients>
		<width>384</width>
		<height>640</height>
		<x>330</x>
		<y>5</y>
		<maximizable>false</maximizable>
		<minimizable>false</minimizable>
		<resizable>false</resizable>
	</initialWindow>
	<icon>
		<image40x40>icn_ios/icn_20_2x.png</image40x40>
		<image60x60>icn_ios/icn_20_3x.png</image60x60>
		<image29x29>icn_ios/icn_29_1x.png</image29x29>
		<image58x58>icn_ios/icn_29_2x.png</image58x58>
		<image87x87>icn_ios/icn_29_3x.png</image87x87>	
		<image80x80>icn_ios/icn_40_2x.png</image80x80>
		<image120x120>icn_ios/icn_40_3x.png</image120x120>	
		<image180x180>icn_ios/icn_60_3x.png</image180x180>
		<image76x76>icn_ios/icn_76_1x.png</image76x76>
		<image152x152>icn_ios/icn_76_2x.png</image152x152>	
		<image167x167>icn_ios/icn_83.5_2x.png</image167x167>
	</icon>
	<customUpdateUI>false</customUpdateUI>
	<allowBrowserInvocation>false</allowBrowserInvocation>
	
	<iPhone>    
		<InfoAdditions>
			<![CDATA[		
		<key>UIBackgroundModes</key>
		<array>
			<string>audio</string>
		</array>
		
		<key>UIDeviceFamily</key>
		<array>
			<string>1</string>
			<string>2</string>
		</array>		
		<key>MinimumOSVersion</key>
      <string>9.0</string>
		<key>NSAppTransportSecurity</key>
		<dict>
			<key>NSAllowsArbitraryLoads</key>
			<true/>
		</dict>
			
			]]>
		</InfoAdditions>
		<requestedDisplayResolution>high</requestedDisplayResolution>
	</iPhone>

	<extensions>
		<extensionID>com.distriqt.Core</extensionID>
		<extensionID>com.distriqt.Share</extensionID>		
		<extensionID>com.distriqt.MediaPlayer</extensionID>		
	</extensions>

</application>
```
  
  - And I compile Dev and Dis files (iOS) for Apple by using OpenSSL in CMD with a command like this:

`E:\sdk33.1.1.476\bin\adt -package -target ipa-test -useLegacyAOT yes -provisioning-profile qh_dev.mobileprovision -storetype pkcs12 -keystore win_dev.p12 -storepass xxxxxxxx QH7.3.23_Dev.ipa QuranHakim_ios.xml QH.swf icn_ios dta -extdir ane Default~iphone.png Default-375w-667h@2x~iphone.png Default-414w-736h@3x~iphone.png Default-568h@2x~iphone.png Default-812h@3x~iphone.png Default-Portrait@2x.png Default-Portrait@2x~ipad.png Default-Portrait~ipad.png Default-Portrait-1112h@2x.png Default@2x~iphone.png Assets.car`

**Question 1**: by executing the above code, I get this error (new for me):

```
Exception in thread "main" java.lang.IllegalAccessError: class com.adobe.air.ipa.MachoSigner (in unnamed module @0x5cee5251) cannot access class sun.security.x509.X500Name (in module java.base) because module java.base does not export sun.security.x509 to unnamed module @0x5cee5251
        at com.adobe.air.ipa.MachoSigner.isDistributionCert(MachoSigner.java)
        ....
```

**Question 2**: Is it possible/do I need to add SDK33 in Flash CS6? It [does not add](https://pasteboard.co/K0zFpPY.png) for me. So inevitably, I use the latest version of SDK32 provided by Adobe inside Flash (32.0.0.144) and then copy the generated SWF file to the folder that I generate the **ipa** (and **apk** / **exe**) files using OpenSSL and Harman's latest SDK 33.

https://pasteboard.co/K0zCW1V.png

**Question 3**: Is this the best way to compile the iOS file on Windows based on my work environment?

Thank you very much for your help! have a nice day

**Comment by ajwfrost** [comment, marked as answer]
Hi

For Q1: this implies that you've got an incompatible Java runtime being used, that doesn't contain those (sun.security) classes. Can you please ensure you've got something like OpenJDK, it works with versions 8 through 12 at least from what I've seen.

Q2 -> I suspect this is because the latest AIR SDK requires Java 8 to compile the Android libraries. Flash CS6 maybe contains a JRE that's older.. you get the same thing with the distributions of Flash Builder, and can update the JRE within those so potentially you could do the same thing with CS6?

Q3 .. if you're not able to sort out Flash CS6 to use the latest SDK, then yes this would be a possible workaround. There are always the two separate steps: (a) compile the SWF; (b) package up into the target format. Using one SDK for compiling and the other for the packaging is fine, the only limitation is that the compiler won't give you the latest AS3 language/API updates; if you're not using them, then that's not a problem at all. You could always copy the newer airglobal.swc into the older SDK if you needed that... (but please back that up first before overwriting it!)

thanks

**Reply by Divaryab** [reply]
Thank you so much for your time and complete reply! After some hours I will have access to my computer, will follow your advice and come back here with the results. Have a wonderful day!

**Reply by Divaryab** [reply]
Ok, first I should mention the 3rd answer was really what I was looking for. So I have no more worry about making SWF inside Flash with AIR 32 and then compile the iOS / APK separately using SDK 33. Nice!

Regarding Java, I re-downloaded the latest version ([16.0.1](https://www.oracle.com/java/technologies/javase-jdk16-downloads.html)) and re-installed it, but same error: 
https://pasteboard.co/K0E71Vl.png

And the same problem exist when I want to compile APK:
`E:\sdk33.1.1.476\bin\adt -package -target apk-captive-runtime -storetype pkcs12 -keystore QuranHakim_apk.p12 -storepass xxxxxx QH7.3_32bit.apk QuranHakim_apk.xml QH.swf icn_apk dta -extdir ane`

so I will do more investigation and share the results here.

Thanks again, will be back

**Reply by Divaryab** [reply]
Wow, I didn't know I need to downgrade Java from 16 to 8! Thanks to your hint and the answer in this link:
https://forum.starling-framework.org/d/22994-sdk33-ios-and-building-on-windows-issue

Problem solved!

**Reply by ajwfrost** [reply]
I'm using OpenJDK 14.0.2 and it's working for me - although I do also have 1.8.0.201 on my machine so it's possible it's picking that up.

We need to remove this dependency on the internal private JRE classes! (if nothing else, we get hundreds of warnings when we build it..)

Have just added some more thoughts into the Starling forums too..
